[JSC] Move FTLAbstractHeap related classes to B3

Constellation opened the pull request:
#### 0eea29cd306f792105f66858379ce2a850dbfd9a
<pre>
[JSC] Move FTLAbstractHeap related classes to B3
<a href="https://bugs.webkit.org/show_bug.cgi?id=299327">https://bugs.webkit.org/show_bug.cgi?id=299327</a>
<a href="https://rdar.apple.com/161127432">rdar://161127432</a>

Reviewed by Yijia Huang and Justin Michaud.

This patch moves FTL::AbstractHeap, FTL::AbstractHeapRepository, and
FTL::TypedPointer to B3. This is the first step making B3 more high-level
understanding JS and Wasm well in the frontend (and lowering them in the
pipeline). And long term goal is removing FTL SSA and use B3 SSA
instead, but anyway, this is the simple &quot;moving component from FTL to
B3&quot; change.

This change allows OMG to use these concepts, allowing OMG to do CSE for
memory access in wasm.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/JavaScriptCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3AbstractHeap.cpp: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeap.cpp.
(JSC::B3::IndexedAbstractHeap::atSlow):
(JSC::B3::IndexedAbstractHeap::initialize):
* Source/JavaScriptCore/b3/B3AbstractHeap.h: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeap.h.
(JSC::B3::AbstractHeap::initialize):
(JSC::B3::AbstractHeap::range const):
(JSC::B3::AbstractHeap::mutability const):
(JSC::B3::IndexedAbstractHeap::offset const):
(JSC::B3::IndexedAbstractHeap::elementSize const):
* Source/JavaScriptCore/b3/B3AbstractHeapRepository.cpp: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp.
(JSC::B3::AbstractHeapRepository::AbstractHeapRepository):
(JSC::B3::AbstractHeapRepository::decorateMemory):
(JSC::B3::AbstractHeapRepository::decorateCCallRead):
(JSC::B3::AbstractHeapRepository::decorateCCallWrite):
(JSC::B3::AbstractHeapRepository::decoratePatchpointRead):
(JSC::B3::AbstractHeapRepository::decoratePatchpointWrite):
(JSC::B3::AbstractHeapRepository::decorateFenceRead):
(JSC::B3::AbstractHeapRepository::decorateFenceWrite):
(JSC::B3::AbstractHeapRepository::decorateFencedAccess):
(JSC::B3::AbstractHeapRepository::computeRangesAndDecorateInstructions):
* Source/JavaScriptCore/b3/B3AbstractHeapRepository.h: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h.
(JSC::B3::AbstractHeapRepository::forIndexingType):
(JSC::B3::AbstractHeapRepository::HeapForValue::HeapForValue):
* Source/JavaScriptCore/b3/B3TypedPointer.h: Renamed from Source/JavaScriptCore/ftl/FTLTypedPointer.h.
(JSC::B3::TypedPointer::TypedPointer):
(JSC::B3::TypedPointer::value const):
* Source/JavaScriptCore/ftl/FTLAbbreviatedTypes.h:
* Source/JavaScriptCore/ftl/FTLCommonValues.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::entrySwitch):
(JSC::FTL::Output::baseIndex):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::parseAndCompileOMG):

Canonical link: <a href="https://commits.webkit.org/300394@main">https://commits.webkit.org/300394@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd9a1096f2b79230cba7831228305cdba6f8c89a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122429 "8 style errors") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/42135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/32821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/129021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/42854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/50728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/129021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125381 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/42854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/32821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/129021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/42854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/32821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/72514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/114577 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/42854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/32821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/131757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120955 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/49370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/50728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/131757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/49744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/32821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/131757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/32821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19350 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/49228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/54972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151114 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/48695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/38651 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/50377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2a35aa84293a1faa7c2310939e77ac3d302efad7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51148

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3c36dfc4c523d107af2cf562f070fe86fef22ef4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51148

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5883842088287682ae6729c47d986875e69fc5b3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51148

Constellation committed:
[JSC] Move FTLAbstractHeap related classes to B3
https://bugs.webkit.org/show_bug.cgi?id=299327
rdar://161127432

Reviewed by Yijia Huang and Justin Michaud.

This patch moves FTL::AbstractHeap, FTL::AbstractHeapRepository, and
FTL::TypedPointer to B3. This is the first step making B3 more high-level
understanding JS and Wasm well in the frontend (and lowering them in the
pipeline). And long term goal is removing FTL SSA and use B3 SSA
instead, but anyway, this is the simple "moving component from FTL to
B3" change.

This change allows OMG to use these concepts, allowing OMG to do CSE for
memory access in wasm.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/JavaScriptCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3AbstractHeap.cpp: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeap.cpp.
(JSC::B3::IndexedAbstractHeap::atSlow):
(JSC::B3::IndexedAbstractHeap::initialize):
* Source/JavaScriptCore/b3/B3AbstractHeap.h: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeap.h.
(JSC::B3::AbstractHeap::initialize):
(JSC::B3::AbstractHeap::range const):
(JSC::B3::AbstractHeap::mutability const):
(JSC::B3::IndexedAbstractHeap::offset const):
(JSC::B3::IndexedAbstractHeap::elementSize const):
* Source/JavaScriptCore/b3/B3AbstractHeapRepository.cpp: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp.
(JSC::B3::AbstractHeapRepository::AbstractHeapRepository):
(JSC::B3::AbstractHeapRepository::decorateMemory):
(JSC::B3::AbstractHeapRepository::decorateCCallRead):
(JSC::B3::AbstractHeapRepository::decorateCCallWrite):
(JSC::B3::AbstractHeapRepository::decoratePatchpointRead):
(JSC::B3::AbstractHeapRepository::decoratePatchpointWrite):
(JSC::B3::AbstractHeapRepository::decorateFenceRead):
(JSC::B3::AbstractHeapRepository::decorateFenceWrite):
(JSC::B3::AbstractHeapRepository::decorateFencedAccess):
(JSC::B3::AbstractHeapRepository::computeRangesAndDecorateInstructions):
* Source/JavaScriptCore/b3/B3AbstractHeapRepository.h: Renamed from Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h.
(JSC::B3::AbstractHeapRepository::forIndexingType):
(JSC::B3::AbstractHeapRepository::HeapForValue::HeapForValue):
* Source/JavaScriptCore/b3/B3TypedPointer.h: Renamed from Source/JavaScriptCore/ftl/FTLTypedPointer.h.
(JSC::B3::TypedPointer::TypedPointer):
(JSC::B3::TypedPointer::value const):
* Source/JavaScriptCore/ftl/FTLAbbreviatedTypes.h:
* Source/JavaScriptCore/ftl/FTLCommonValues.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::entrySwitch):
(JSC::FTL::Output::baseIndex):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::parseAndCompileOMG):

Canonical link: https://commits.webkit.org/300394@main
